### PR TITLE
zcash: 5.3.0 -> 5.4.2

### PR DESCRIPTION
**Pull request by teor2345**
###### Description of changes

- Add new native dependency git and build dependency tl-expected
- Update source and cargo hashes

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by centromere** [comment]
@teor2345, have you seen https://github.com/TartanLlama/expected/pull/117?

**Comment by teor2345** [comment]
> @teor2345, have you seen [TartanLlama/expected#117](https://github.com/TartanLlama/expected/pull/117)?

I had not, thanks for letting me know.

I don't think it's a blocker for this PR, is there some other action you think we should take?

**Comment by centromere** [comment]
It's true that zcash may build without including the latest changes to tl-expected. However, in the spirit of correctness I think we should wait until @daira's changes (which were approved and merged) are available in nixpkgs.

I got the ball rolling with #216551. Hopefully it can be merged soon.

**Comment by teor2345** [comment]
> It's true that zcash may build without including the latest changes to tl-expected. However, in the spirit of correctness I think we should wait until @daira's changes (which were approved and merged) are available in nixpkgs.
> 
> I got the ball rolling with #216551. Hopefully it can be merged soon.

This patch is already applied as part of the zcash build, so there's no need to wait:
https://github.com/zcash/zcash/pull/6138/files

Also, the current zcashd version in nixpkgs has reached its expiry halt and no longer runs.

**Comment by centromere** [comment]
@teor2345 Those Makefiles aren't used when building with Nix because, when invoked, `make` would access the Internet and violate reproducibility.

**Comment by teor2345** [comment]
> @teor2345 Those Makefiles aren't used when building with Nix because, when invoked, `make` would access the Internet and violate reproducibility.

Fair enough.

But we need this new zcash version urgently, because the current version has reached end of life and no longer runs.

**Comment by centromere** [comment]
The aforementioned changes were merged to master a few minutes ago, so you should be clear to rebase.

**Review by centromere — APPROVED** [review]

File: pkgs/applications/blockchains/zcash/default.nix
Comment (centromere) [inline pkgs/applications/blockchains/zcash/default.nix:32]: @teor2345 I just noticed... `tl-expected` is _only_ a header file. As such, it is a build time dependency, not a runtime dependency. Would it be more appropriate to list it as a `nativeBuildInput` rather than a `buildInput`?

**Review by centromere — CHANGES_REQUESTED** [review]

File: pkgs/applications/blockchains/zcash/default.nix
Comment (teor2345) [reply pkgs/applications/blockchains/zcash/default.nix]: Fixed in 83f85a7

**Comment by teor2345** [comment]
And that's rebased now. Happy to let you handle it from now on!

**Review by centromere — APPROVED** [review]

File: pkgs/applications/blockchains/zcash/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/zcash/default.nix:25]: tl-expected looks like it goes to buildInputs.

Comment (centromere) [reply pkgs/applications/blockchains/zcash/default.nix:25]: `tl-expected` only contains a single header file. It is not a runtime dependency.

Comment (teor2345) [reply pkgs/applications/blockchains/zcash/default.nix:25]: I'm going to be away for a while, feel free to make whatever edits you need here. (Maintainer access is on.)

**Comment by teor2345** [comment]
@centromere are you able to make whatever changes are needed here?

`zcashd` has released a security update recently, we should probably use that:
https://github.com/zcash/zcash/releases/tag/v5.4.2

File: pkgs/applications/blockchains/zcash/default.nix
Comment (teor2345) [reply pkgs/applications/blockchains/zcash/default.nix:25]: I fixed this in the latest PR update.

File: pkgs/applications/blockchains/zcash/default.nix
Comment (centromere) [reply pkgs/applications/blockchains/zcash/default.nix:32]: > Remove header-only dependency tl-expected from nativeBuildInputs

Why was this action taken?

**Review by centromere — CHANGES_REQUESTED** [review]

File: pkgs/applications/blockchains/zcash/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/blockchains/zcash/default.nix:32]: It is easier if we stuff all libraries into buildInputs. Also in the future the package might change and be a single header file plus a script which then could be for the wrong architecture.

**Comment by SuperSandro2000** [comment]
@ofborg build zcash

**Pull request merged**
